js/layout-javascript/news-feed-code.js

```diff
@@ -379,10 +387,18 @@ DynamicList.prototype.attachObservers = function() {
 
         // find the element to expand and expand it
         if (_this.allowClick) {
-          $el.parents('.new-news-feed-list-container').addClass('hidden');
           _this.$container.find('.dynamic-list-add-item').addClass('hidden');
 
-          _this.showDetails(entryId);
+          _this.showDetails(entryId).then(function() {
+            setTimeout(function() {
+              _this.$closeButton.focus();
+              _this.$detailsContent.focusin(function() {
```

Comment:
Can you explain why the `.focusin()` and `.focusout()` event listeners are necessary?

Reply:
When the focus disappears from the item in the popup, we shift the focus to the close buttonю.  But if the next focus element is the popup element, we undo the transition of the focus to the close button.
 
focusout =>  _this.$closeButton.focus();
focusin => cancellation of operation focusout if the next focus element is the popup element

Reply:
> When the focus disappears from the item in the popup, we shift the focus to the close button.

Very good idea. It wasn't discussed in our scenario planning, but this sounds like the right thing to do. It would be good to have this discussed and clarified first on Jira so QA also knows about it first. cc @romanyosyfiv @inna-bieshulia 

`_this.focusCloseButton(false)` doesn't make sense though. I took a look at the `focusCloseButton` function and if the parameter is `false`, the function simply does nothing. It doesn't look like any event listener is needed for `focusin`.

Reply:
@tonytlwu _this.focusCloseButton(false) cancels focus on the close button when the next focus element is in the popup. For this I use debounce.

Reply:
@OleksandrLavrynenko This is where the `focusCloseButton()` is defined.

```js
DynamicList.prototype.focusCloseButton = _.debounce(function(isAction) {
  if (isAction) {
    this.$closeButton.focus();
  }
}, 200);
```

The debounced function is:

```js
function(isAction) {
  if (isAction) {
    this.$closeButton.focus();
  }
}
```

`_this.focusCloseButton(false)` calls the debounced function with `false` as its parameter.

```js
```js
function(isAction) {
  if (isAction) {
    this.$closeButton.focus();
  }
}
```

...does nothing if `isAction` is `false`.

Have I misunderstood something? It seems that this line does nothing.

Reply:
**@tonytlwu** Yes, it does nothing but undo the previous action when true is passed. When `focusout `is triggered, the `_this.focusCloseButton (true)` function should shift the focus to the close button, but when `focusin `is triggered, the `_this.focusCloseButton (false)` function cancels this action. Thus, when the focus is on the last element and we press the Tab key, `focusout ` is started, and since the next element does not exist, this action is not canceled, and the focus goes to the close button.

Reply:
@OleksandrLavrynenko When `focusin` is triggered, how does `_this.focusCloseButton(false)` cancel the previous focus shift if `_this.focusCloseButton(false)` programmatically does nothing?

It looks like you can achieve the same result by just taking out the `focusin` event listener.

Reply:
**@tonytlwu** No. If the `focusin `is off then the `focusout `will always assign focus to the close button

Reply:
@OleksandrLavrynenko I see what you tried to do after trying out the event listeners myself. It looks like `focusin` and `focusout` is occurring on all focusable elements inside the container, and not just the container. This means both `focusout` and `focusin` (in that specific order) are always being triggered.

Based on [this article](https://stackoverflow.com/questions/15418358/how-to-detect-when-a-container-loses-focus), I created the following solution which is cleaner and clearer at describing what's happening.

https://jsfiddle.net/tonytlwu/b3so9pLq/

Let me know if you disagree.

Reply:
@tonytlwu That's cool! Thanks! I forgot about relatedTarget